### PR TITLE
Add cmake-lint

### LINKED ISSUE #1132: Add cmake-lint
## What

Add `cmake-lint` to build

## Why

Ensure a consistent style in CMake files. This is especially important as we start to attract outside collaborators.

## Notes

We have prior art in [TraitGen](https://github.com/OpenAssetIO/OpenAssetIO-TraitGen/blob/591863d593d21d988da2bcd8b8207467e7582d85/tests/generators/cpp/cmake/StaticAnalyzers.cmake#L90)

## Acceptance Criteria

- Determine applicability to other repos and log issues so we have consistent approach across projects (don't implement)
- `cmake-lint` enabled via a `cmake` option following the pattern of `cpplint` enabled in the `lint` preset.
- linter checks must match results of available formatting tools.

### PR DESCRIPTION
**Pull request by feltech**
## Description

Closes #1132.

Following the pattern of other linters, e.g. `cpplint`, add an optional build target that executes `cmake-lint`, and wire that target as a dependency of all targets configured using the `openassetio_set_default_target_properties` function.

Add to `linter-requirements.txt` and enable the cmake-lint check in linting CMake presets, so that it is installed and runs on CI (and well-behaved local dev machines).

- [ ] ~~I have updated the release notes.~~
- [ ] ~~I have updated all relevant user documentation.~~

## Reviewer Notes

Do we need release notes for this?  Its kinda internal.

## Test Instructions

`drop!` commit illustrates failure.

### CODE REVIEW

File: CMakeLists.txt [hunk 1]
```diff
@@ -333,6 +339,7 @@ endif ()
 message(STATUS "Linter: clang-tidy                 = ${OPENASSETIO_ENABLE_CLANG_TIDY} [${OPENASSETIO_CLANGTIDY_EXE}]")
 message(STATUS "Linter: cpplint                    = ${OPENASSETIO_ENABLE_CPPLINT} [${OPENASSETIO_CPPLINT_EXE}]")
 message(STATUS "Linter: clang-format               = ${OPENASSETIO_ENABLE_CLANG_FORMAT} [${OPENASSETIO_CLANGFORMAT_EXE}]")
+message(STATUS "Linter: cmake-lint                 = ${OPENASSETIO_ENABLE_CMAKE_LINT} [${OPENASSETIO_CMAKELINT_EXE}]")
```
Comment (foundrytom) [inline CMakeLists.txt:342]: [Fix] `BUILDING.md` needs updating.

File: cmake/StaticAnalyzers.cmake [hunk 2]
```diff
@@ -85,3 +85,36 @@ macro(enable_clang_Format)
...
+# Linter configuration is in cmake-format.yaml.
+macro(enable_cmake_lint)
+    find_program(OPENASSETIO_CMAKELINT_EXE NAMES cmake-lint)
+    if (OPENASSETIO_CMAKELINT_EXE)
+        file(
+            GLOB_RECURSE _sources
+            LIST_DIRECTORIES false
+            CONFIGURE_DEPENDS # Ensure we re-scan if files change.
+            ${PROJECT_SOURCE_DIR}/cmake/*.cmake
+            ${PROJECT_SOURCE_DIR}/src/CMakeLists.txt
```
Comment (foundrytom) [inline cmake/StaticAnalyzers.cmake:102]: [question] (Sorry CMake noob) Do we need to trigger the rescan if the other nested `CMakeLists.txt` files too?

**Review by foundrytom — CHANGES_REQUESTED** [review]

File: cmake/StaticAnalyzers.cmake (hunk 2 above)
Comment (feltech) [reply cmake/StaticAnalyzers.cmake:102]: Not sure what you mean by trigger the rescan. GLOB_RECURSE is weird in that the last component of a file path (i.e. the file name) is searched for recursively under the given prefix (i.e. everything up to the file name).  So the nested `CMakeLists.txt` are discovered.  I'll break one of those deliberately to demonstrate.

Comment (feltech) [reply cmake/StaticAnalyzers.cmake:102]: Added a violation, see https://github.com/OpenAssetIO/OpenAssetIO/actions/runs/6889451328/job/18740464053?pr=1208

Comment (foundrytom) [reply cmake/StaticAnalyzers.cmake:102]: Oh that is confusing, even without any wild cards? Their docs dont make that particularly clear.... but makes sense, thanks.

File: CMakeLists.txt (hunk 1 above)
Comment (feltech) [reply CMakeLists.txt:342]: Well remembered, updated in 6335d2cc

**Review by foundrytom — APPROVED** [review]

**Review by elliotcmorris — APPROVED** [review]

**Pull request merged**